### ISSUE TITLE
Concepts without equality

### ISSUE THREAD

**Issue by stefjoosten** [AmpersandTarski/Ampersand#722]
# Symptoms
There are atoms without equality. Examples are blobs and videos. (Strictly speaking, real numbers are an example too. Real numbers are typically implemented by floating point arithmetic with limited precision, for which equality can be implemented as bitwise equal).
# Problem
Ampersand assumes that equality is defined for all atoms. This assumption being false, we can expect funny behaviour.

**Comment by stefjoosten** [comment]
# Analysis
(written with @sjcjoosten )
We distinguish concept based on whether:
1. equality is implemented for the concept
2. enumeration is implemented for the concept

An example of a concept for which both are implemented could be ``User``.
An example of a concept for which equality is not, but enumeration is implemented, could be a concept represented as ``BLOB``.
An example of a concept for which enumeration is not implemented, could be the natural numbers and the reals (for which equality could and can't be implemented respectively).

To decide what can and what cannot be implemented, we look at a function that needs to be computed when showing ``VIEW``s and interfaces:
We look at whether, for an expression ``E``, we can compute a function ``f`` that takes an atom from the source of ``E`` as its argument, and produces the set of elements in the image of ``E`` to which that atom is mapped, formally:

``f(x) = {z | (x,z) \in E}``

**Comment by stefjoosten** [comment]
# Solution
In Ampersand, we want the type checker to verify the properties of the various concepts, without the user giving any explicit syntax.

type | equality | enumerable
------------ | ------------- | -------
Alphanumeric | yes | no
BigAlphanumeric | no | no
HugeAlphanumeric  | no | no
Password  | yes | no
Binary | no | no
BigBinary  | no | no
HugeBinary  | no | no
Date  | yes | no
DateTime  | yes | no
Boolean | yes | yes
Integer  | yes | no
Float  | yes | no
Object  | yes | yes
TypeOfOne  | yes | yes

A concept without representation type is treated as yes/yes, which retains the classic Ampersand semantics in scripts without ``REPRESENT``-statement.

**Comment by stefjoosten** [comment]
# Solution 2
Consider
~~~
CLASSIFY A ISA B
~~~
The equality of ``A`` inherits from ``B``.
``A`` is at least as enumerable as ``B``. So: if ``B`` is enumerable, ``A`` must be enumerable too.
If we implement this, we'll start by making all user defined concepts enumerable so we don't have to introduce new syntax. Later, we can introduce syntax for the user to decide about enumerability of ``A``.

**Comment by stefjoosten** [comment]
# Solution 3

**Comment by stefjoosten** [comment]
# Consequences will be:
1. In an interface or a view, a concept without equality can occur only at the right hand side of an expression.

**Comment by stefjoosten** [comment]
# Consequences will be:
1. In an interface or a view, a concept without equality can occur only at the right hand side of an expression.
2. The existence of ``f`` for every expression means that the fields in interfaces and views are computable. This prevents the erratic behaviour mentioned in the problem statement.
3. To detect violations by means of ``f`` opens the possibility to detect violations selectively. We might be able to show only the violations related to the current session, rather than all violations of everyone (including all other users), which is what happens currently. That, however is a different issue.

**Comment by RieksJ** [comment]
I do not quite understand the problem at hand. Equality is properly defined (as bitwise equality). Apparently, @stefjoosten has something else in his mind. To me this sounds as if the representations that we use for entities that are known to exist within some CONTEXT (which are the bitrows for which we have a bitwise equality) are confused with the entities themselves. 

@stefjoosten and @sjcjoosten: can you clarify (i.e. edit the first entry of this issue)?

**Comment by sjcjoosten** [comment]
I edited the first paragraph. Does this help?

**Comment by RieksJ** [comment]
No. I still do not see what the problem is. If equality is implemented but not needed - just don't use it. What is the _problem_?

**Comment by sjcjoosten** [comment]
That equality is not implemented in these cases.
Can you rephrase the introduction such that this becomes clear?

**Comment by RieksJ** [comment]
I could perhaps if I would know what the problem is you're trying to address. 
If you have not implemented equality and you do not want it, there is no problem.
If you have implemented equality and you do not want it, there is no problem (you ignore it/do not use it).
If you cannot define equality and you don't want it, there is no problem.
What's the problem?

**Comment by Michiel-s** [comment]
For the prototypes, equality also depends on the  implementation. E.g. in a mysql database, equality for floating point variables does not work in all cases. This is something we should take into account.

**Comment by Michiel-s** [comment]
Sorry @hanjoosten, I wanted to quote reply, but instead edited your comment above. Picked the wrong button in the Github app.
